### ISSUE TITLE
Test `injection_queue_depth_multi_thread` is flaky

### ISSUE THREAD

**Issue by Darksonn** [tokio-rs/tokio#6847]
This test has been observed to fail in CI:

https://github.com/tokio-rs/tokio/blob/83e922f051e341e4d69d04c7a8ef1050c19cb0f8/tokio/tests/rt_unstable_metrics.rs#L642-L668

To close this issue, figure out why it is failing and fix it.

**Comment by jofas** [comment]
I presume it's these two recent jobs that timed out

* https://github.com/tokio-rs/tokio/actions/runs/10820423666/job/30020516601

* https://github.com/tokio-rs/tokio/actions/runs/10718157570/job/29719516533

that have been affected by the flakiness of `injection_queue_depth_multi_thread`? I'd assume so because if this assert

https://github.com/tokio-rs/tokio/blob/83e922f051e341e4d69d04c7a8ef1050c19cb0f8/tokio/tests/rt_unstable_metrics.rs#L663

fails, the main thread panics and we never get to synchronise on `barrier2`

https://github.com/tokio-rs/tokio/blob/83e922f051e341e4d69d04c7a8ef1050c19cb0f8/tokio/tests/rt_unstable_metrics.rs#L667

here, causing the test to run forever (or until CI times out).

**Comment by Darksonn** [comment]
Yes. Good point with the assert. That explains why it times out instead of failing normally.

**Comment by jofas** [comment]
I wonder if a first step[^1] would be to convert the assert into an if-statement where we'd panic only after calling `barrier2.wait()`. Then we'd get better diagnostics but more importantly wouldn't time out CI any more.

[^1]: In case this isn't a quick fix

**Comment by Darksonn** [comment]
Making it fail instead of hanging would be nice. I submitted a PR for that.

**Comment by jofas** [comment]
Interestingly enough, after I pulled your changes, the test still stalls instead of failing. It seems like I was wrong about the `assert!` causing a panic which makes us never reach the second barrier. 

At least I built this highly professional bash script&mdash;which I'm going to share here in case someone else wants to use it&mdash;that finally allowed me to trigger the flakiness on my local x86_64 Linux system:

```bash
run_test() {
  cargo test \
    --all-features \
    --test rt_metrics \
    injection_queue_depth_multi_thread
}

i=0
while run_test; do
  let i++
  echo -e "$i: \e[32m☑\e[0m"
done
```

Now that I can litter my fork with debug print statements, I'll investigate some more :slightly_smiling_face:

**Comment by jofas** [comment]
Uhm, isn't the flakiness just a mundane case of accidentally [blocking](https://doc.rust-lang.org/std/sync/struct.Barrier.html#method.wait) the runtime before it can schedule the second task?

**Comment by Darksonn** [comment]
The runtime isn't supposed to get blocked by this. If there is an idle worker thread and a runnable task, the worker thread must pick up a runnable task. The only exception is the LIFO slot which is not relevant here.

**Comment by jofas** [comment]
Could there be some weirdness around parked threads? The deadlocked test [parks](https://github.com/tokio-rs/tokio/blob/82628b8a7828507617a430e6875f1e25d5c211ef/tokio/src/runtime/scheduler/multi_thread/worker.rs#L748) the second worker before it gets the second task in its work queue and does not unpark it again on my machine.

**Comment by jofas** [comment]
I'd love to give it a try for sure. This is quite exciting. @Darksonn may I ask you questions in case I get stuck on this?

**Comment by Darksonn** [comment]
Yes, feel free to send questions my way. A loom test would be a good start.
